modules/core/src/test/java/org/apache/ignite/internal/util/DistributedProcessTest.java

```diff
@@ -0,0 +1,162 @@
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        startGrids(SRV_NODES);
+
+        startClientGrid(SRV_NODES);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() {
+        stopAllGrids();
+    }
+
+    /** */
+    @Test
+    public void testBackgroundExecFailureHandled() throws Exception {
+        checkDistributedProcess((n, latch) ->
+            new DistributedProcess<>(grid(n).context(), TEST_PROCESS,
+                req -> runAsync(() -> {
+                    failOnNode(n);  // Fails processing request in a spawned thread.
+
+                    return 0;
+                }),
+                (id, res, err) -> {
+                    if (failCln)
+                        assertEquals(SRV_NODES, res.values().size());
+                    else {
+                        assertEquals(SRV_NODES - 1, res.values().size());
+                        assertEquals(1, err.size());
+                        assertTrue(err.get(grid(1).localNode().id()) instanceof AssertionError);
+                    }
+
+                    latch.countDown();
+                }));
+    }
+
+    /** */
+    @Test
+    public void testExecFailureHandled() throws Exception {
+        checkDistributedProcess((n, latch) ->
+            new DistributedProcess<>(grid(n).context(), TEST_PROCESS,
+                req -> {
+                    failOnNode(n);  // Fails processing request in the discovery thread.
+
+                    return new GridFinishedFuture<>(0);
+                },
+                (id, res, err) -> {
+                    if (failCln)
+                        assertEquals(SRV_NODES, res.values().size());
+                    else {
+                        assertEquals(SRV_NODES - 1, res.values().size());
+                        assertEquals(1, err.size());
+                        assertTrue(err.get(grid(1).localNode().id()) instanceof AssertionError);
+                    }
+
+                    latch.countDown();
+                }));
+    }
+
+    /** */
+    @Test
+    public void testFinishFailureHandled() throws Exception {
+        checkDistributedProcess((n, latch) ->
+            new DistributedProcess<>(grid(n).context(), TEST_PROCESS,
+                req -> new GridFinishedFuture<>(0),
+                (uuid, res, err) -> {
+                    assertEquals(SRV_NODES, res.values().size());
+                    latch.countDown();
+
+                    failOnNode(n);
+                }));
+    }
+
+    /** */
+    private void checkDistributedProcess(
+        BiFunction<Integer, CountDownLatch, DistributedProcess<Integer, Integer>> processFactory
+    ) throws Exception {
+        DistributedProcess<Integer, Integer> process = null;
+
+        CountDownLatch latch = new CountDownLatch(SRV_NODES + 1);
+
+        for (int n = SRV_NODES; n >= 0; n--)
+            process = processFactory.apply(n, latch);
```

Comment:
It seems that a bunch of processes is created here but only one of them actually is started